### PR TITLE
feat(51000) - Flag Deprecation Plan

### LINKED ISSUE #51000: Flag Deprecation Plan
## TypeScript Flag Deprecation Plan

> (This is a proposal, but written in terms of a blog post for clarity)

As we think about the long-term future of TypeScript and its place in the JavaScript ecosystem, it's clear that many things TypeScript 0.8 users in 2012 needed aren't things that TypeScript users in 2024 are going to need. For example, in 2012, Internet Explorer 8 (which only supports ECMAScript 3) had enough marketshare to warrant serious compatibility concern. Today, thankfully, it doesn't.

While we still consider long-term language stability to be an absolutely paramount concern, having a well-defined deprecation policy in place also makes it easier for us to be *slightly* less risk-averse when adding new features to the language.

To that end, starting with TypeScript 5.0, we will begin the process of removing certain flags, features, and/or behaviors which we think the language or ecosystem have largely evolved past the need for.

### Schedule

The deprecation schedule is based on our existing `version += 0.1` versioning cadence, aligned with ones-place-incrementing versions for clarity.

|Release / Phase|Description|
|----|------|
|TypeScript 5.0 - Phase 1|❔ Deprecation warnings introduced|
|TypeScript 5.5 - Phase 2|⚠ Flags start doing nothing|
|TypeScript 6.0 - Phase 3|❌ Flags removed entirely|

The cycle repeats again with TypeScript 6.0 also introducing whatever new deprecation warnings deemed appropriate at that time.

#### Phase 1: ❔ Deprecation Warnings

In phase 1 (e.g. TypeScript 5.0), using a deprecated flags will produce a commandline error:
```
TS9998: Flag 'keyofStringsOnly' is deprecated and will stop functioning in TypeScript 5.5. Specify 'ignoreDeprecations: "5.0"' to silence this error
```

As hinted to in the error, you can supply a new flag value to silence this error:
```
    ignoreDeprecations: "5.0"
```

All other behavior is unchanged.

This gives people an immediate warning of impending deprecation, but won't block them from upgrading to TypeScript 5.0.

#### Phase 2: ⚠ Flags Start Doing Nothing 

In phase 2 (e.g. TypeScript 5.5), deprecated flags can still be *specified* in tsconfig / commandline settings, but no longer have any effect. This allows users to have `tsconfig.json` files that work on both sides of the deprecation cycle (as you might have in a monorepo with a "shared" tsconfig settings file).

#### Phase 3: ❌ Flags Removed

In phase 3 (e.g. TypeScript 6.0), it is an error to specify a deprecated flag.

At this point, it also becomes illegal to have `ignoreDeprecations: "5.0`, since this setting doesn't do anything (it's illegal to specify those flags in the first place, and they've done nothing for 5 versions).

In 6.0, the only legal value of `ignoreDeprecations` will then be `"6.0"`, per its Phase 1.

## 5.0 Deprecation Candidates

This list is provisional; please leave comments if you are depending on these in a serious way.

Removed entirely:

 * `charset` - already deprecated since approximately the beginning of time
 * `noImplicitUseStrict` - non-strict-mode code is extremely rare in modern JS
 * `noStrictGenericChecks` - this flag is broadly a bad idea, and improvements to the type system have rendered its original use cases largely moot
 * `out` - already deprecated since approximately the beginning of time (use `outFile` instead)
 * `keyofStringsOnly` - according to a rumor, we only wanted to ship this for three releases (2.9, 3.0, 3.1) when we added this!
 * `suppressExcessPropertyErrors` - improvements in not performing subtype reduction have rendered its original use cases largely moot
 * `suppressImplicitAnyIndexErrors` - the use cases here have been supplanted by better recognition of literal-typed keys
 * `noFallthroughCasesInSwitch` - style concern; use a linter if this is not allowed in your coding style

Removed possible values:

 * `target` - remove `ES3` - no known extant software runs ES3 (this will be even more true in 2 years), and we don't have `lib` support for it anyway
 * `module` - remove `umd`, `system`, and `amd` - the ecosystem is rapidly converging, and better downlevelers exist for each of these should you still need them by 2024

### PR DESCRIPTION
**Pull request by a-tarasyuk**
Fixes #51000

### CODE REVIEW

**Comment by a-tarasyuk** [comment]
Many tests don't set `target` explicitly and rely on the `ES3`, therefore, the PR contains many changes related to tests

https://github.com/microsoft/TypeScript/blob/896b9228710941f875294ebee2da998d500c8614/src/harness/harnessIO.ts#L393

https://github.com/microsoft/TypeScript/blob/896b9228710941f875294ebee2da998d500c8614/src/compiler/utilities.ts#L6361-L6366

Is it acceptable to set globally `ignoreDeprecations` for tests?

**Comment by Jack-Works** [comment]
> Many tests don't set target explicitly and rely on the ES3, therefore, the PR contains many changes related to tests

can you create a new pr to explicitly set tests target to es5?

**Comment by a-tarasyuk** [comment]
> can you create a new pr to explicitly set tests target to es5?

I can, however many changes will remain due to the difference in output between _es3_ and _es5_.

**Comment by Jack-Works** [comment]
> > can you create a new pr to explicitly set tests target to es5?
> 
> I can, however many changes will remain due to the difference in output between _es3_ and _es5_.

but that PR won't need to be reviewed clearly

**Comment by Jack-Works** [comment]
I suggest to deprecate `--experimentalDecorators` and add a new flag `--legacyDecorators` to migrate

**Comment by RyanCavanaugh** [comment]
Some notes from reviewing this:
 * Thanks for taking this on!
 * I'll send a PR shortly to change the default target to ES5. The test impact isn't *too* bad but it's big indeed.
 * In general, I think this code is "too maintanable" as it were. We'll need to update these things every 2.5 years or so, so simply hardcoding the disallowed values altogether should be simpler to read IMO. The "phases" as we have them don't need to be manifested in the code; when 5.5 comes around we'll (hopefully) simply be deleting values from enums. Hope that makes sense and can make this a bit shorter.

**Comment by a-tarasyuk** [comment]
> In general, I think this code is "too maintanable" as it were. We'll need to update these things every 2.5 years or so, so simply hardcoding the disallowed values altogether should be simpler to read IMO. The "phases" as we have them don't need to be manifested in the code; when 5.5 comes around we'll (hopefully) simply be deleting values from enums. Hope that makes sense and can make this a bit shorter.

I've removed the parse versions in favor of the hardcoded versions and left only the type required for `CompilerOptions`. Maybe you have ideas for further simplification...

> "too maintainable"

😂 - this is the first time I've seen such a polite synonym for _"these changes don't meet expectations"._

File: tests/baselines/reference/deprecatedCompilerOptions4.errors.txt [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+/foo/tsconfig.json(10,9): error TS5096: Flag 'out' is deprecated and will stop functioning in TypeScript 6.0. Specify 'ignoreDeprecations: "5.5"' to silence this error.
+/foo/tsconfig.json(11,31): error TS5098: Invalid value for '--ignoreDeprecations'.
+
+
+==== /foo/tsconfig.json (9 errors) ====
+    {
+        "compilerOptions": {
+            "target": "ES3",
+                      ~~~~~
+!!! error TS5096: Flag 'ES3' is deprecated and will stop functioning in TypeScript 6.0. Specify 'ignoreDeprecations: "5.5"' to silence this error.
```
Comment (RyanCavanaugh) [inline tests/baselines/reference/deprecatedCompilerOptions4.errors.txt:17]: Something weird is going on here; the only allowed `ignoreDeprecations` value is `"5.0"` and these features will cease to function in 5.5, not 6.0

**Comment by RyanCavanaugh** [comment]
I think the above comment is the only one I have left - LGTM otherwise. Great work!

File: tests/baselines/reference/es3-jsx-react-native.types [hunk 2]
```diff
@@ -4,8 +4,8 @@ const React: any = null;
 >null : null
 
 const elem = <div></div>;
->elem : error
-><div></div> : error
+>elem : any
+><div></div> : any
```
Comment (andrewbranch) [inline tests/baselines/reference/es3-jsx-react-native.types:8]: Why did the printout of some of these ES3 error types change? Also, how was there an error type but no error baseline before? I feel pretty sure this isn’t a problem but I’m totally confused by it.

File: src/compiler/diagnosticMessages.json [hunk 3]
```diff
@@ -4245,6 +4245,18 @@
         "category": "Error",
         "code": 5100
     },
+    "Flag '{0}' is deprecated and will stop functioning in TypeScript {1}. Specify 'ignoreDeprecations: \"{2}\"' to silence this error.": {
+        "category": "Error",
+        "code": 5101
+    },
+    "Flag '{0}' is deprecated, please remove it from your configuration.": {
```
Comment (andrewbranch) [inline src/compiler/diagnosticMessages.json:4252]:
```suggestion
    "Flag '{0}' is deprecated; please remove it from your configuration.": {
```

or 

```suggestion
    "Flag '{0}' is deprecated. Please remove it from your configuration.": {
```

?

File: tests/baselines/reference/es3-jsx-react-native.types (hunk 2 above)
Comment (RyanCavanaugh) [reply tests/baselines/reference/es3-jsx-react-native.types:8]: Guessing: I believe configuration errors are supposed to effectively halt all future steps, so the checker is in an impossible-to-reach situation in this case

File: src/compiler/diagnosticMessages.json (hunk 3 above)
Comment (a-tarasyuk) [reply src/compiler/diagnosticMessages.json]: I've changed to 
> Flag '{0}' is deprecated. Please remove it from your configuration.

**Review by andrewbranch — APPROVED** [review]
Thanks @a-tarasyuk!

**Comment by RyanCavanaugh** [comment]
🎉!

**Pull request merged**